Document fusion methods: weighted RRF, DBSF, FormulaQuery (#2357)
qdrant/landing_page: qdrant-landing/content/documentation/search/hybrid-queries.md

Reviewer: If we change the first snippet to already use the RRF query, maybe we should change this into something like:

```suggestion
To set the constant $k$ in the formula:
```

{{< code-snippet path="/documentation/headless/snippets/query-points/hybrid-rrf-weights/" >}}

Weights are a configuration choice, not something you can tune arbitrarily. The most reliable way to set them is by testing on your data.

- **With an eval set (queries paired with known-relevant docs):** split your eval queries in two. Try different weights on the first half, then measure on the second half. Measuring on the same queries you tuned on inflates the result. The [Choosing a Fusion Method notebook](https://githubtocolab.com/qdrant/examples/blob/master/fusion-methods/Choosing_a_Fusion_Method.ipynb) provides a reusable `tune_rrf_weights` grid-search helper you can adapt to a train/val split.
- **Without an eval set:** leave weights at the default `(1.0, 1.0)`. Hand-tuned weights without measurement are unlikely to beat the default reliably.

Retune when your retrievers change (new embedding model, new chunking), when your corpus drifts substantially, or on a fixed cadence with a fresh eval sample.

### Distribution-Based Score Fusion (DBSF)

_Available as of v1.11.0_

> <a href=https://medium.com/plain-simple-software/distribution-based-score-fusion-dbsf-a-new-approach-to-vector-search-ranking-f87c37488b18 target="_blank">DBSF</a> keeps the raw scores from each query but normalizes their distributions before combining. For each retriever's returned set, it computes the mean $\mu$ and sample standard deviation $\sigma$, then linearly remaps every score using the 3-sigma extremes as endpoints:

Reviewer 2: maybe keep _normalization_ keyword?
```suggestion
<a href=https://medium.com/plain-simple-software/distribution-based-score-fusion-dbsf-a-new-approach-to-vector-search-ranking-f87c37488b18 target="_blank">DBSF</a> keeps the raw scores from each query but normalizes their distributions before combining. For each retriever's returned set, it computes the mean $\mu$ and sample standard deviation $\sigma$, then normalizes every score using the 3-sigma extremes as endpoints:
```

$$ \hat{s} = \frac{s - (\mu - 3\sigma)}{6\sigma} $$

Normalized scores are summed across retrievers. Different score magnitudes no longer matter because each retriever contributes on the same comparable range.

<aside role="status"><code>dbsf</code> is stateless and computes its normalization limits from each query's returned points, not from all the scores it has seen. Scores are <strong>not</strong> clipped to [0, 1]; values outside the 3-sigma range remain outside it after the remap. If all returned scores are identical (or only one point is returned), DBSF emits <code>0.5</code> rather than dividing by zero.</aside>

> DBSF is a reasonable choice when you trust your retrievers' raw scores to carry magnitude information. On well-calibrated retrievers DBSF can outperform tuned weighted RRF; on others weighted RRF wins. Neither dominates the other in general, so use your eval set to choose between them. Two caveats apply: the statistics come from the prefetch top-k (a small sample), and a single dominant outlier in that top-k can skew normalization for that query. Increase the prefetch `limit` if you see unstable rankings.

Reviewer 2: > Two caveats apply: the statistics come from the prefetch top-k (a small sample), and a single dominant outlier in that top-k can skew normalization for that query. Increase the prefetch `limit` if you see unstable rankings.

Nice addition :+1:

_Available as of v1.17.0_

By default, each query is assigned an equal weight. In reality, one retriever is often stronger than the other for a given workload. For example, a dense retriever may dominate on natural-language queries, while BM25 may win on identifier-heavy ones. Assigning equal weight to both can let the weaker retriever drag down results. To address this, you can assign greater weight to rankers that perform well on your evaluation set.

The `rrf` query allows you to configure relative weights for each of the prefetches. For example, if you have two prefetches and assign a weight of 3.0 to the first and 1.0 to the second, a document ranked third in the first query scores the same as a document ranked first in the second query. In the case of non-overlapping result sets, these weights return three results from the first set for every one result from the second set.

Weights should be provided as an array of numbers, where each weight is applied to the corresponding prefetch in the order they are defined. The number of weights must match the number of prefetches.

{{< code-snippet path="/documentation/headless/snippets/query-points/hybrid-rrf-weights/" >}}

> Weights are a hyperparameter, not a free knob. A held-out eval is the most defensible way to set them.

Reviewer: I think most of our users won't know what a hyperparameter or held-out eval is. How about something like this:
```suggestion
Weights are a configuration choice, not something you can tune arbitrarily. The most reliable way to set them is by testing on your data.
```

Normalized scores are summed across retrievers. Different score magnitudes no longer matter because each retriever contributes on the same comparable range.

<aside role="status"><code>dbsf</code> is stateless and computes its normalization limits from each query's returned points, not from all the scores it has seen. Scores are <strong>not</strong> clipped to [0, 1]; values outside the 3-sigma range remain outside it after the remap. If all returned scores are identical (or only one point is returned), DBSF emits <code>0.5</code> rather than dividing by zero.</aside>

DBSF is a reasonable choice when you trust your retrievers' raw scores to carry magnitude information. On well-calibrated retrievers DBSF can outperform tuned weighted RRF; on others weighted RRF wins. Neither dominates the other in general, so use your eval set to choose between them. Two caveats apply: the statistics come from the prefetch top-k (a small sample), and a single dominant outlier in that top-k can skew normalization for that query. Increase the prefetch `limit` if you see unstable rankings.

{{< code-snippet path="/documentation/headless/snippets/query-points/hybrid-dbsf/" >}}

> ### Custom Fusion with FormulaQuery

Reviewer: I was wondering if this belongs here, or in the "Multi-Stage Queries" section. We already have rescoring examples there, and this feels more like a rescoring than a fusion method?

Author: Moved to Multi-Stage Queries as "Custom Scoring with `FormulaQuery`" in a42055e. The "Choosing a Fusion Method" table now forward-links to it for the rescoring use case.

_Available as of v1.14.0_

> `FormulaQuery` lets you write the combining expression explicitly, using scores from prefetches (`$score`), payload fields, and built-in helpers like `ExpDecayExpression` or `GaussDecayExpression`. It is **not** a replacement for tuned weighted RRF. Writing `0.7 * $score[0] + 0.3 * $score[1]` over raw retriever scores reintroduces the same scale problem that breaks naive linear fusion. If the prefetches are themselves `rrf` or `dbsf`, the scores are already on comparable scales and a weighted formula sum works.

Reviewer: ```suggestion
`FormulaQuery` lets you write the combining expression explicitly, using scores from prefetches (`$score`), payload fields, and built-in helpers like `ExpDecayExpression` or `GaussDecayExpression`. It is **not** a replacement for tuned weighted RRF. Writing `0.7 * \$score[0] + 0.3 * \$score[1]` over raw retriever scores reintroduces the same scale problem that breaks naive linear fusion. If the prefetches are themselves `rrf` or `dbsf`, the scores are already on comparable scales and a weighted formula sum works.
```
Otherwise this doesn't render correctly

> `FormulaQuery` lets you write the combining expression explicitly, using scores from prefetches (`$score`), payload fields, and built-in helpers like `ExpDecayExpression` or `GaussDecayExpression`. It is **not** a replacement for tuned weighted RRF. Writing `0.7 * $score[0] + 0.3 * $score[1]` over raw retriever scores reintroduces the same scale problem that breaks naive linear fusion. If the prefetches are themselves `rrf` or `dbsf`, the scores are already on comparable scales and a weighted formula sum works.

Reviewer: How about using "The Formula Query" instead of `FormulaQuery`?

Author: Heading is now "Custom Scoring with `FormulaQuery`" and the body opens "A `FormulaQuery` lets you compose..." (a42055e). Kept the code form since it is the SDK class name; the article gives it the prose feel.
